### PR TITLE
feat(tasks): ensure blocking task relations block completion

### LINKED ISSUE #2194: Blocked tasks can be marked completed
### Pre-submission checklist

- [x] I have searched for existing open or closed issue reports with the same problem.

### Description

I expect that if I added a blocking relationship between tasks, the "Blocked By" task should not be allowed to be marked completed without the Blocking task being marked completed first, but this is not enforced.

### Vikunja Version

v1.0.0-rc4

### Browser and version

_No response_

### Can you reproduce the bug on the Vikunja demo site?

Yes

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by HarshPatel5940**
Resolves #2194 

This pull request introduces logic to enforce task blocking: tasks that are blocked by other incomplete tasks cannot be marked as complete. It adds a new error type for this scenario, updates the task update process to check for blockers, and provides comprehensive tests to ensure the correct enforcement of blocking relationships.

**Task blocking enforcement:**

* Added a check in `Task.updateSingleTask` to prevent marking a task as complete if any blocking tasks are not yet completed, returning a new `ErrTaskIsBlocked` error with details about the blockers.
* Implemented the helper function `getActiveBlockingRelations` to retrieve all incomplete tasks that are blocking a given task.

**Error handling improvements:**

* Introduced the `ErrTaskIsBlocked` error type, including HTTP error mapping, error code, and human-readable details about blocking tasks; also added a function to check for this error type.

**Testing enhancements:**

* Added a comprehensive set of tests in `TestTaskBlockingEnforcement` to verify that blocked tasks cannot be completed until all blockers are done, and to check error details and edge cases.

### CODE REVIEW

**Comment by HarshPatel5940** [comment]
working on the concern ux

**Comment by HarshPatel5940** [comment]
@kolaente :) ping pong

**Comment by HarshPatel5940** [comment]
@kolaente is this something you require me to resolve conflicts and work?

or you never want this feature and are gonna close? this way i save any more efforts in this PR if so

**Comment by kolaente** [comment]
Hey sorry for the silence! I have a bunch of things on my plate right now with my day job and other Vikunja things.

I would like to have this feature and think it makes sense to have it, just need to set aside some time to review it properly.

Don't worry about the conflicts, we'll resolve these shortly before merging.

**Comment by HarshPatel5940** [comment]
> Hey sorry for the silence! I have a bunch of things on my plate right now with my day job and other Vikunja things.
> 
> I would like to have this feature and think it makes sense to have it, just need to set aside some time to review it properly.
> 
> Don't worry about the conflicts, we'll resolve these shortly before merging.

Make sense! Thank you for the reply and it's understandable!

Will resolve the conflicts 🫂

**Comment by kolaente** [comment]
Please check the failing frontend lint though

File: frontend/src/i18n/lang/en.json [hunk 1]
```diff
@@ -1457,7 +1458,8 @@
     "13002": "The provided link share password is invalid.",
     "13003": "The provided link share token is invalid.",
     "14001": "The provided api token is invalid.",
-    "14002": "The permission {permission} of group {group} is invalid."
+    "14002": "The permission {permission} of group {group} is invalid.",
+    "20001": "This task is blocked by incomplete tasks and cannot be marked as done."
```
Comment (kolaente) [inline frontend/src/i18n/lang/en.json]: Why two times the same message?

File: frontend/src/components/tasks/partials/SingleTaskInProject.vue [hunk 2]
```diff
@@ -13,12 +13,12 @@
 			@keyup.enter="openTaskDetail"
 		>
 			<span
-				v-tooltip="!canMarkAsDone ? $t('task.readOnlyCheckbox') : ''"
+				v-tooltip="!canMarkAsDone ? $t('task.readOnlyCheckbox') : isBlockedByIncomplete ? $t('task.blockedCheckbox') : ''"
```
Comment (kolaente) [inline frontend/src/components/tasks/partials/SingleTaskInProject.vue]: please put this into a computed, nesting tenary operators gets too complex pretty quickly

File: frontend/src/views/tasks/TaskDetailView.vue [hunk 3]
```diff
@@ -1106,7 +1113,14 @@ async function saveTask(
 		currentTask.endDate = currentTask.dueDate
 	}
 
-	const updatedTask = await taskStore.update(currentTask) // TODO: markraw ?
+	let updatedTask: ITask
```
Comment (kolaente) [inline frontend/src/views/tasks/TaskDetailView.vue:1124]: not sure how much this is in scope for this PR, but it feels like the logic is duplicated between here and SingleTaskInProject. Does it make sense to move the common parts to a composable and call in both places?

File: pkg/models/tasks.go [hunk 4]
```diff
@@ -2026,3 +2050,34 @@ func triggerTaskUpdatedEventForTaskID(s *xorm.Session, auth web.Auth, taskID int
...
+
+	// Extract IDs of blocking tasks
+	blockingTaskIDs := make([]int64, 0, len(relations))
+	for _, rel := range relations {
+		blockingTaskIDs = append(blockingTaskIDs, rel.OtherTaskID)
+	}
+
+	// Get all blocking tasks that are NOT complete
+	blockingTasks := []*Task{}
+	err = s.In("id", blockingTaskIDs).Where("done = ?", false).Find(&blockingTasks)
```
Comment (kolaente) [inline pkg/models/tasks.go]: Let's do the whole thing in one sql statement (one sub query that fetches the tasks, the other one that checks their status)

**Review by kolaente — CHANGES_REQUESTED** [review]

File: frontend/src/i18n/lang/en.json (hunk 1 above)
Comment (HarshPatel5940) [reply frontend/src/i18n/lang/en.json]: same msg?

File: frontend/src/components/tasks/partials/SingleTaskInProject.vue (hunk 2 above)
Comment (HarshPatel5940) [reply frontend/src/components/tasks/partials/SingleTaskInProject.vue]: i agree

File: pkg/models/tasks.go (hunk 4 above)
Comment (HarshPatel5940) [reply pkg/models/tasks.go]: noted

File: frontend/src/views/tasks/TaskDetailView.vue (hunk 3 above)
Comment (HarshPatel5940) [reply frontend/src/views/tasks/TaskDetailView.vue:1124]: hmm true, will make an composable and call

**Comment by HarshPatel5940** [comment]
will clear the comments, lints and merge and update shortly.

**Comment by HarshPatel5940** [comment]
the comments have been addressed.

File: frontend/src/i18n/lang/en.json (hunk 1 above)
Comment (kolaente) [reply frontend/src/i18n/lang/en.json]: It's the pretty much the same as in line 944

File: pkg/models/tasks.go [hunk 5]
```diff
@@ -1255,6 +1256,29 @@ func (t *Task) updateSingleTask(s *xorm.Session, a web.Auth, fields []string) (e
...
+			// Build informative error message with blocker details
+			blockingInfo := make([]string, 0, len(blockingTasks))
+			blockingIDs := make([]int64, 0, len(blockingTasks))
+			for _, blocker := range blockingTasks {
+				blockingInfo = append(blockingInfo, fmt.Sprintf("%s (ID: %d)", blocker.Title, blocker.ID))
+				blockingIDs = append(blockingIDs, blocker.ID)
+			}
+			return ErrTaskIsBlocked{
+				TaskID:           t.ID,
+				BlockingTaskIDs:  blockingIDs,
```
Comment (kolaente) [inline pkg/models/tasks.go]: Just add the full list of blocking tasks, building a human readable error message is a concern of the client consuming the API.

**Review by kolaente — CHANGES_REQUESTED** [review]

File: pkg/models/tasks.go (hunk 5 above)
Comment (HarshPatel5940) [reply pkg/models/tasks.go]: hmm okay

**Comment by HarshPatel5940** [comment]
@kolaente i did some cleanup, I know you asked me to return the list and frontend just builts msg for it.

Until i realise frontend already has the knowledge of the tasks and can check **locally for just relationships. So i did that.**

This saves an api call i presume and thats good right? 🤔 tho i am still worried of some edge case scenarios maybe you help me on confirm this.

**Comment by kolaente** [comment]
it might be possible in the kanban or gantt view that the subtasks are not loaded for certain filters. But as you said, that's an edge case. You could do a fallback thing: If the tasks are not present locally, do a fetch.
IMHO just returning the full tasks in the error and then building the proper errror string in the frontend should handle all edge cases. Since we're already fetching the full tasks during the relation blocking check, we don't save a roundtrip by not returning them, and the implementation would be more robust.

**Comment by HarshPatel5940** [comment]
Ahhhh make sense. Will do it and reply back.

**Comment by HarshPatel5940** [comment]
@kolaente updated as requested.

sorry for the delayed fix

File: pkg/models/error.go [hunk 6]
```diff
@@ -2627,6 +2627,45 @@ func (err ErrTimeEntryEndBeforeStart) HTTPError() web.HTTPError {
...
+	_, ok := err.(ErrTaskIsBlocked)
+	return ok
+}
+
+func (err ErrTaskIsBlocked) Error() string {
+	return fmt.Sprintf("task %d is blocked by %d task(s)", err.TaskID, len(err.BlockingTasks))
+}
+
+// ErrCodeTaskIsBlocked holds the unique world-error code of this error
+const ErrCodeTaskIsBlocked = 20001
```
Comment (kolaente) [inline pkg/models/error.go:2717]: If this is errors 20_000, it should be placed after 19_000 in the order. 

More importantly, I don't think we should add a new class of errors just for this, just add it to the task errors.

**Review by kolaente — CHANGES_REQUESTED** [review]

**Comment by HarshPatel5940** [comment]
whoof

**Comment by HarshPatel5940** [comment]
got it

**Comment by HarshPatel5940** [comment]
remaking PR ig
